internal/interface/crawler/crawler.go

```diff
@@ -0,0 +1,185 @@
+package crawler
+
+import (
+	"bytes"
+	"database/sql"
+	"fmt"
+	"github.com/PuerkitoBio/goquery"
+	"github.com/gocolly/colly/v2"
+	"github.com/mp-hl-2021/unarXiv/internal/domain"
+	"github.com/mp-hl-2021/unarXiv/internal/domain/model"
+	"github.com/mp-hl-2021/unarXiv/internal/domain/repository"
+	"github.com/mp-hl-2021/unarXiv/internal/interface/utils"
+	"strings"
+	"time"
+)
+
+const kURLBacklogSize = 1000
+
+var (
+	ErrNoConfigs     = fmt.Errorf("no configurations found")
+	ErrExpectedHref  = fmt.Errorf("expected attribute href but it wasn't found")
+	ErrTooShortAbsId = fmt.Errorf("too short absId")
+	ErrEmptyTitle    = fmt.Errorf("parsed empty title")
+	ErrEmptyAuthors  = fmt.Errorf("parsed empty authors")
+)
+
+type Crawler struct {
+	db           *sql.DB
+	articlesRepo repository.ArticleRepo
+}
+
+func NewCrawler(db *sql.DB, articlesRepo repository.ArticleRepo) *Crawler {
+	return &Crawler{db: db, articlesRepo: articlesRepo}
+}
+
+type Configuration struct {
+	Id                  int
+	DesiredArticleCount int
+	RootURL             string
+}
+
+func (c *Crawler) GetConfiguration() (Configuration, error) {
+	rows, err := c.db.Query("SELECT * FROM CrawlerConfig;")
+	if err != nil {
+		return Configuration{}, err
+	}
+	defer rows.Close()
+	for rows.Next() {
+		cfg := Configuration{}
+		err := rows.Scan(&cfg.Id, &cfg.DesiredArticleCount, &cfg.RootURL)
+		return cfg, err
+	}
+	return Configuration{}, ErrNoConfigs
+}
+
+func (c *Crawler) getArticlesCount() (int, error) {
+	rows, err := c.db.Query("SELECT COUNT(*) FROM Articles;")
+	if err != nil {
+		return 0, err
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var cnt int
+		err = rows.Scan(&cnt)
+		return cnt, err
+	}
+	return 0, fmt.Errorf("unexpected query result")
+}
+
+func (c *Crawler) upsertArticle(article model.Article) (bool, error) {
```

Comment:
I failed miserably to wrap my head around this logic..

Reply:
made it more readable